transports/webrtc/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+//! - The offerer sends this SDP-encoded message to the answerer. The medium through which this
+//!   exchange is done is out of scope of the ICE protocol.
+//! - The answerer then finds its own candidates, and generates an answer, again in the SDP format.
+//!   This answer is sent back to the offerer.
+//! - Each agent then tries to connect to the remote's candidates.
+//!
+//! We pretend to send the offer to the remote agent (the target of the connection), then pretend
+//! that it has found a valid IP address for itself (i.e. a candidate), then pretend that the SDP
+//! answer containing this candidate has been sent back. This will cause the offerer to execute
+//! step 4: try to connect to the remote's candidate.
+//!
+//! ## 1 vs N data channels
+//!
+//! The SDP message generated by the offerer contains the list of so-called "media streams" that it
+//! wants to open. In our specific use-case, we configure the transport to always request one data
+//! stream.
+//!
+//! For this prototype, we have choosen to implement a single data channel and multiplex on top
+//! because doing N channels will force us to rewrite a significant part of smoldot, which isn't
+//! ready at all to have encryption and multiplexing handled "externally".
+//!
+//! ## TCP or UDP
+//!
+//! WebRTC by itself doesn't hardcode any specific protocol for these media streams. Instead, it is
+//! the SDP message of the offerer that specifies which protocol to use. In our use case, one data
+//! stream, we know that the offerer will always request either TCP+DTLS+SCTP, or UDP+DTLS+SCTP.
+//!
+//! ## DTLS+SCTP
+//!
+//! RFCs: 8841, 8832
+//!
+//! In both cases (TCP or UDP), the next layer is DTLS. DTLS is similar to the well-known TLS
+//! protocol, except that it doesn't guarantee ordering of delivery (as this is instead provided by
+//! the SCTP layer on top of DTLS). In other words, once the TCP or UDP connection is established,
+//! the browser will try to perform a DTLS handshake.
+//!
+//! During the ICE negotiation, each agent must include in its SDP packet a hash of the self-signed
+//! certificate that it will use during the DTLS handshake. In our use-case, where we try to
+//! hand-crate the SDP answer generated by the remote, this is problematic. One way to solve this
+//! is to make the hash a part of the remote's multiaddr.
+//!
+//! ## PeerId
+//!
+//! Ideally, we would just add a field in the certificate that contains a signature of the
+//! certificate made using the PeerId. But you can't do that, as it's a chicken and egg problem:
+//! you can't modify the certificate anymore after you've created a signature of it. We could find
+//! some semi-hacky system where the signature is made against the certificate but with this field
+//! removed, but that's very hacky as well.
+//!
+//! For now we'll start an encryption protocol handshake on top of the single data channel. Once
+//! this handshake has been successful, we stop it and just use DTLS.
```

Comment:
I also don't think we need anything related to Noise or ECDH or anything of the sort. I don't even remember why doing a partial Noise handshake was on the table, but it was probably partly because the code already exists.

All we need in reality is to pass a signature of the certificate made using the PeerId.

Reply:
Haven't put enough thought into this, nor do I have the expertise to make an informed decision.

> All we need in reality is to pass a signature of the certificate made using the PeerId.

The simpler the better.